homeassistant/components/bluecurrent/__init__.py

```diff
@@ -0,0 +1,244 @@
+    """Set all Blue Current entities to unavailable."""
+    registry = entity_registry.async_get(hass)
+    entries = entity_registry.async_entries_for_config_entry(registry, config_id)
+
+    for entry in entries:
+        entry.write_unavailable_state(hass)
+
+
+class Connector:
+    """Define a class that connects to the Blue Current websocket API."""
+
+    charge_points: dict[str, dict] = {}
+    grid: dict[str, Any] = {}
+
+    def __init__(
+        self, hass: HomeAssistant, config: ConfigEntry, client: Client
+    ) -> None:
+        """Initialize."""
+        self._config: ConfigEntry = config
+        self._hass: HomeAssistant = hass
+        self.client: Client = client
+
+    async def connect(self, token: str) -> None:
+        """Register on_data and connect to the websocket."""
+        await self.client.connect(token)
+
+    async def on_data(self, message: dict) -> None:
+        """Handle received data."""
+
+        async def handle_charge_points(data: list) -> None:
+            """Loop over the charge points and get their data."""
+            for entry in data:
+                evse_id = entry[EVSE_ID]
+                model = entry[MODEL_TYPE]
+                self.add_charge_point(evse_id, model)
+                await self.get_charge_point_data(evse_id)
+            await self.client.get_grid_status(data[0][EVSE_ID])
+
+        object_name: str = message[OBJECT]
+
+        if DATA in message:
+            data: dict | list = message[DATA]
+
+        # gets charge point ids
+        if object_name == CHARGE_POINTS:
+            assert isinstance(data, list)
+            await handle_charge_points(data)
+
+        # gets charge point key / values
+        elif object_name in VALUE_TYPES:
+            assert isinstance(data, dict)
+            evse_id = data.pop(EVSE_ID)
+            self.update_charge_point(evse_id, data)
+
+        # gets grid key / values
+        elif object_name == GRID_STATUS:
+            assert isinstance(data, dict)
+            self.grid = data
+            self.dispatch_signal()
+
+    async def get_charge_point_data(self, evse_id: str) -> None:
+        """Get all the data of a charge point."""
+        await self.client.get_status(evse_id)
+
+    def add_charge_point(self, evse_id: str, model: str) -> None:
+        """Add a charge point to charge_points."""
+        self.charge_points[evse_id] = {MODEL_TYPE: model}
+
+    def update_charge_point(self, evse_id: str, data: dict) -> None:
+        """Update the charge point data."""
+        for key in data:
+            self.charge_points[evse_id][key] = data[key]
+        self.dispatch_signal(evse_id)
+
+    def dispatch_signal(self, evse_id: str | None = None) -> None:
+        """Dispatch a signal."""
+        if evse_id is not None:
+            async_dispatcher_send(self._hass, f"{DOMAIN}_value_update_{evse_id}")
+        else:
+            async_dispatcher_send(self._hass, f"{DOMAIN}_grid_update")
```

Comment:
Feels a bit weird for this function to have two signals to send depending on whether the argument is None. I'd split these to `dispatch_value_update_signal(evse_id)` and `dispatch_grid_update_signal()`.

Reply:
Split into separate methods.